### DISCUSSION TITLE
Use string union type values for schema

### DISCUSSION THREAD

**Discussion by bennettdams** [colinhacks/zod#768] [Q&A] [answered]
String unions should be preferred over enums [for several reasons](https://fettblog.eu/tidy-typescript-avoid-enums/).

Is there a way to create a Zod schema from them?

Here's an example where I want to create a schema for input validation when updating a user:

**types.ts**
```ts
export type Role = "USER" | "ADMIN"

export type User = {
  username: string;
  role: Role
}

```
**SomeComponent.tsx**
```ts
// a schema to do input validation when updating a user
const userInputSchema = z.object({
  username: z.string().min(2, "Minimum 2 characters"),
  password: z.string().min(6, "Minimum 6 characters"),
  // this is what I don't know
  role: ?
});
```

There's a [recent discussions](https://github.com/colinhacks/zod/discussions/760) that shows how to do it the other way around: Create the type from a Zod schema which defines the literals. 
For my example, this would look like this:

```ts
const roleSchema = z.union([z.literal('USER'), z.literal('ADMIN')])
type Role = z.infer<typeof roleSchema>

type User = {
  username: string;
  role: Role
}

// a schema to do input validation when updating a user
const userInputSchema = z.object({
  username: z.string().min(2, "Minimum 2 characters"),
  password: z.string().min(6, "Minimum 6 characters"),
  role: roleSchema
});
```

This would work, but I don't want to take this route, because everytime I use the type `Role` somewhere else to create a new schema, I would need to use the `roleSchema`. So I'm switching between using my `Role` type to describe data and the `roleSchema` for a new schema.

Is there an easier way to just use the string union values from a type like `Role`?

**Comment by JacobWeisenburger** [comment, marked as answer]
> Is there an easier way to just use the string union values from a type like Role?

What are you trying to use the "string union values" for? Because `Role` is defined in typescript, you can only use it in typescript. Javascript doesn't know `Role` exists, because types are removed when you compile to javascript.
___
> This would work, but I don't want to take this route, because everytime I use the type Role somewhere else to create a new schema, I would need to use the roleSchema. So I'm switching between using my Role type to describe data and the roleSchema for a new schema.

I guess I'm not sure what's wrong with using Role to make a type and using roleSchema to make a schema. But if you are only wanting to use one of them and not both, then you could just not use Role and infer the type of User from userInputSchema. See below.

```ts
const roleSchema = z.union( [
    z.literal( 'USER' ),
    z.literal( 'ADMIN' ),
] )

const userInputSchema = z.object( {
    username: z.string().min( 2, "Minimum 2 characters" ),
    password: z.string().min( 6, "Minimum 6 characters" ),
    role: roleSchema
} )
type User = z.infer<typeof userInputSchema>
```

If this isn't a good solution for you, please explain what the problem is with using Role to make a type and using roleSchema to make a schema

**Reply by bennettdams** [reply]
> What are you trying to use the "string union values" for? Because Role is defined in typescript, you can only use it in typescript. Javascript doesn't know Role exists, because types are removed when you compile to javascript.

Aren't string union types for this kind of operation special, because they're "represented" as strings in JavaScript as well?

```ts
type Role = "USER" | "ADMIN"

function doSomething(role: Role): void {
  if (role === "ADMIN") {
    // ...
  } else if (role === "USER") {
    // ...
  }
  ...
}

```
---


In the original example above, I want to use the type `Role` to describe my data, everywhere it is used.
I also need the role in some spaces to verify user input, as described above.
This means I definitely need the type `Role`, but don't necessarily need the `roleSchema`:

Coming back to my example, I e.g. need the role in another schema like so:
```ts
const userInputSchema = z.object({
  username: z.string().min(2, "Minimum 2 characters"),
  password: z.string().min(6, "Minimum 6 characters"),
  // this is what I don't know
  role: ?
});
```
Of course it works using the `roleSchema` here like you and I described above, I just thought it would be nice to just use the type `Role` on the fly when needing it in another schema. I was asking if there's a way to somehow generate/infer the schema directly from an existing string union type, something like this:
```ts
type Role = "USER" | "ADMIN"

const userInputSchema = z.object({
  username: z.string().min(2, "Minimum 2 characters"),
  password: z.string().min(6, "Minimum 6 characters"),
  // infer `Role` type, because it is a static string union
  // this is equal to `z.union([z.literal('USER'), z.literal('ADMIN')])`
  role: z.someMethod<Role>
});
```
What I'm mentioning here is not something you can't do with a schema right now - I just think it's one step less to come to the same result.

**Reply by JacobWeisenburger** [reply]
I think you are getting types and values confused.
- Types are only used by typescript at compile time. Types are removed by the typescript compiler when you compile to javascript.
- Values are used by javascript. Using `typeof` allows typescript to get the type of a value. But there isn't a way to get a value from a type.

```ts
type Role = "USER" | "ADMIN"
// Role is a type and javascript can't use this because it will be removed during compilation

const roleSchema = z.union( [ z.literal( 'USER' ), z.literal( 'ADMIN' ) ] )
// roleSchema is a value and javascript can use this

const userInputSchema = z.object( {
    // ...
    role: Role,
    //    ^^^^
    // 'Role' only refers to a type, but is being used as a value here.

    role: z.someMethod<Role>
    //    ^^^^^^^^^^^^^^^^^^
    // 'z.someMethod<Role>' only refers to a type, but is being used as a value here.

    role: roleSchema,
    // this works because roleSchema is a value
} )

// Role does NOT equal roleSchema
// In fact these can't even be compared, because types don't exist at run time
```

For more info, please check out this video [Fixing TypeScript's Blindspot: Runtime Typechecking](https://www.youtube.com/watch?v=rY_XqfSHock)

**Reply by bennettdams** [reply]
> I think you are getting types and values confused.

I hope I'm well aware of "type" vs. "value", but you're right, I assumed you can use the values of string unions.
If you look at the compiled code, the string values are kept as is:

https://www.typescriptlang.org/play?#code/C4TwDgpgBASg9gG2gXigIgKoGUCiM1QA+6AggCICyAkgHJoCwAUEwMZwB2AzsFAIYAmAWwCW7AFyxEKUpVoNGbLjwCunCACcJ8JFFSZc+JkwBmy9i2DCOUfnCxxBEYAAtRAcwAU6qVqkBKKABvJihQqGFjLyldZD1yajoAxU4pADoEOE80YU4+IVE0PxCwiKidWL1sPEKoZLSMrJyoVQ1CpgBfJiA

![image](https://user-images.githubusercontent.com/29319414/144890327-f0cedb3f-9eac-418a-9341-23212547c246.png)

My idea was that Zod receives the type (a string union), extracts the values (e.g. `"ADMIN"` and `"USER"`) and uses them to check at runtime.
For that reason, before your comment, I thought you can use the values, as they live at runtime as well. I didn't realize that you cannot extract the static string union values from the type, so thanks for the hint.
